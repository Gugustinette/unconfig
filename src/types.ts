import { Arrayable } from '@antfu/utils'

export const defaultExtensions = ['mts', 'cts', 'ts', 'mjs', 'cjs', 'js', 'json', '']

export type ConfigLoaderType = 'bundle' | 'json'

export interface LoadConfigSource<T = any> {
  files: Arrayable<string>

  /**
   * @default ['mts', 'cts', 'ts', 'mjs', 'cjs', 'js', 'json', '']
   */
  extensions?: string[]

  /**
   * Loader for loading config,
   *
   * @default 'auto'
   */
  loader?: ConfigLoaderType | 'auto'

  /**
   * Rewrite the config object
   */
  rewrite?: <F = any>(obj: F, filepath: string, loader: ConfigLoaderType) => Promise<T | undefined> | T | undefined
}

export interface SearchOptions {
  /**
   * Root directory
   *
   * @default process.cwd()
   */
  cwd?: string
}

export interface LoadConfigOptions extends SearchOptions {
  sources: Arrayable<LoadConfigSource>
}

export interface LoadConfigResult<T> {
  config: T
  filepath: string
  mtime: number
}
