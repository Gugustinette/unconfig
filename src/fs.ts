import { dirname, parse, resolve } from 'node:path'
import process from 'node:process'
import { lstat, stat } from '@quansync/fs'
import { quansync } from 'quansync/macro'

export interface FindUpOptions {
  /**
   * @default process.cwd
   */
  cwd?: string
  /**
   * @default path.parse(cwd).root
   */
  stopAt?: string
  /**
   * @default false
   */
  multiple?: boolean
  /**
   * @default true
   */
  allowSymlinks?: boolean
}

const isFile = quansync(async (path: string, allowSymlinks: boolean) => {
  try {
    return (await (allowSymlinks ? stat : lstat)(path)).isFile()
  }
  catch {
    return false
  }
})

export const findUp = quansync(
  async (paths: string[], options: FindUpOptions = {}): Promise<string[]> => {
    const {
      cwd = process.cwd(),
      stopAt = parse(cwd).root,
      multiple = false,
      allowSymlinks = true,
    } = options

    let current = cwd

    const files: string[] = []

    while (current && current !== stopAt) {
      for (const path of paths) {
        const filepath = resolve(current, path)
        if (await isFile(filepath, allowSymlinks)) {
          files.push(filepath)
          if (!multiple)
            return files
        }
      }
      const parent = dirname(current)
      if (parent === current)
        break
      current = parent
    }

    return files
  },
)
