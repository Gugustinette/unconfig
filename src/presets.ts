import { toArray, Arrayable } from '@antfu/utils'
import { LoadConfigSource } from './types'

export interface SourceVitePluginConfigOptions {
  plugins: Arrayable<string>
}

export interface SourceObjectFieldOptions {
  fields: Arrayable<string>
}

export function sourceVitePluginConfig(options: SourceVitePluginConfigOptions): LoadConfigSource {
  const plugins = toArray(options.plugins)
  return {
    files: ['vite.config'],
    async rewrite(obj) {
      const config = await (typeof obj === 'function' ? obj() : obj)
      if (!config)
        return config
      return config.plugins.find((i: any) => plugins.includes(i.name) && i?.api?.config)?.api?.config
    },
  }
}

export function sourceViteConfigFields(options: SourceObjectFieldOptions): LoadConfigSource {
  return {
    files: ['vite.config'],
    ...rewriteFields(options),
  }
}

export function sourceNuxtConfigFields(options: SourceObjectFieldOptions): LoadConfigSource {
  return {
    files: ['nuxt.config'],
    ...rewriteFields(options),
  }
}

export function sourcePackageJsonFields(options: SourceObjectFieldOptions): LoadConfigSource {
  return {
    files: ['package.json'],
    extensions: [],
    loader: 'json',
    ...rewriteFields(options),
  }
}

function rewriteFields(options: SourceObjectFieldOptions): Pick<LoadConfigSource, 'rewrite'> {
  const fields = toArray(options.fields)
  return {
    async rewrite(obj) {
      const config = await (typeof obj === 'function' ? obj() : obj)
      if (!config)
        return config
      for (const field of fields) {
        if (field in config)
          return config[field]
      }
    },
  }
}
