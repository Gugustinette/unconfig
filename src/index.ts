import { promises as fs } from 'fs'
import jiti from 'jiti'
import { toArray } from '@antfu/utils'
import { LoadConfigOptions, LoadConfigResult, LoadConfigSource, SearchOptions, defaultExtensions } from './types'
import { findUp } from './fs'

export * from './types'
export * from './presets'

export async function loadConfig<T>(options?: LoadConfigOptions): Promise<LoadConfigResult<T> | undefined> {
  const sources = toArray(options?.sources || [])
  for (const source of sources) {
    const result = await loadConfigFromSource<T>(source, options)
    if (result)
      return result
  }
}

export async function loadConfigFromSource<T>(source: LoadConfigSource<T>, search: SearchOptions = {}): Promise<LoadConfigResult<T> | undefined> {
  const { extensions = defaultExtensions } = source

  const flatFiles = toArray(source?.files || []).flatMap((file) => {
    return extensions.map(i => i ? `${file}.${i}` : file)
  })

  if (!flatFiles.length)
    return undefined

  const { cwd = process.cwd() } = search
  const files = await findUp(flatFiles, { cwd, stopAt: search.stopAt })

  if (!files.length)
    return undefined

  return await loadConfigFile(files[0], source)
}

async function loadConfigFile<T>(filepath: string, source: LoadConfigSource<T>): Promise<LoadConfigResult<T> | undefined> {
  let config: T | undefined

  let loader = source.loader || 'auto'

  if (loader === 'auto') {
    const content = await fs.readFile(filepath, 'utf-8')
    try {
      config = JSON.parse(content)
      loader = 'json'
    }
    catch {
      loader = 'bundle'
    }
  }
  else {
    loader = 'bundle'
  }

  if (!config) {
    if (loader === 'bundle') {
      config = await jiti(undefined, { interopDefault: true })(filepath)
      if (!config)
        return undefined
    }
    else if (loader === 'json') {
      const content = await fs.readFile(filepath, 'utf-8')
      config = JSON.parse(content)
    }
  }

  if (!config)
    return

  const rewritten = source.rewrite
    ? await source.rewrite(config, filepath, loader)
    : config

  if (!rewritten)
    return undefined

  return {
    config: rewritten,
    sources: [filepath],
  }
}
