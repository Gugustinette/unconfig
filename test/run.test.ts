import { resolve } from 'path'
import { expect, it } from 'vitest'
import { loadConfig } from '../src'
import { sourcePackageJsonFields, sourcePluginFactory } from '../src/presets'

it('load', async () => {
  const cwd = resolve(__dirname, 'fixtures')
  const result = await loadConfig({
    sources: [
      {
        files: 'un.config',
      },
      sourcePackageJsonFields({
        fields: 'un',
      }),
      sourcePluginFactory({
        targetModule: 'stub',
        files: 'rewrite.js',
        extensions: [],
      }),
      sourcePluginFactory({
        targetModule: 'stub',
        files: 'rewrite2.js',
        extensions: [],
      }),
      sourcePluginFactory({
        targetModule: 'stub',
        files: 'params',
        parameters: ['include me', { param2: 'but not me' }],
      }),
    ],
    cwd,
    defaults: {
      defaults: 'default',
      deep: { foo: 'hi' },
    },
    merge: true,
  })

  expect(result.config)
    .toMatchSnapshot()

  expect(result.sources.map(i => i.slice(cwd.length + 1)))
    .toMatchSnapshot('files')
})
