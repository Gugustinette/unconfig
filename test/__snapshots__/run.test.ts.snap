// Vitest Snapshot v1

exports[`load > files 1`] = `5`;

exports[`load 1`] = `
{
  "deep": {
    "foo": "you should get this",
  },
  "defaults": "default",
  "from": "package.json",
  "json": "allo",
  "mjs": "hi",
  "rewrite": "Hi",
  "ts": "hello",
}
`;
